src/python/pants/backend/java/compile/javac.py

```diff
@@ -5,61 +5,168 @@
+        for coarsened_targets in all_coarsened_components
+    )
+    return CoarsenedComponents(
+        tuple(
+            component
+            for _, component in sorted(component_id_to_component.items(), key=lambda x: x[0])
+        )
+    )
+
+
 @rule(level=LogLevel.DEBUG)
 async def compile_java_source(
     coursier: Coursier,
     bash: BashBinary,
     request: CompileJavaSourceRequest,
 ) -> CompiledClassfiles:
-    sources = await Get(
-        SourceFiles,
-        SourceFilesRequest(
-            (request.target.get(Sources),),
-            for_sources_types=(JavaSources,),
-            enable_codegen=True,
+    targets_and_source_files = zip(
+        request.targets,
+        await MultiGet(
+            Get(
+                SourceFiles,
+                SourceFilesRequest(
+                    (target.get(Sources),),
+                    for_sources_types=(JavaSources,),
+                    enable_codegen=True,
+                ),
+            )
+            for target in request.targets
+            if target.has_field(JavaSources)
         ),
     )
-    if sources.snapshot.digest == EMPTY_DIGEST:
+
+    filtered_targets_and_source_files = [
+        (target, sources)
+        for target, sources in targets_and_source_files
+        if sources.snapshot.digest != EMPTY_DIGEST
+    ]
+
+    if not filtered_targets_and_source_files:
         return CompiledClassfiles(digest=EMPTY_DIGEST)
 
-    lockfile, direct_deps = await MultiGet(
-        Get(CoursierResolvedLockfile, CoursierLockfileForTargetRequest(Targets((request.target,)))),
-        Get(Targets, DependenciesRequest(request.target.get(Dependencies))),
+    direct_deps = Addresses(
+        sorted(
+            frozenset(
+                chain.from_iterable(
+                    await MultiGet(
+                        Get(Addresses, DependenciesRequest(target[Dependencies]))
+                        for target, _ in filtered_targets_and_source_files
+                    )
+                )
+            )
+        )
     )
 
-    direct_dependency_classfiles = await MultiGet(
-        Get(CompiledClassfiles, CompileJavaSourceRequest(target=target)) for target in direct_deps
+    canonical_request_component = frozenset(
+        canonical_addresses_for_coarsened_targets(request.targets)
     )
 
+    # It's possible that a target in our input set of coarsened targets has a direct
+    # dependency on another target within the input component (indeed this is guaranteed
+    # to happen if the input component has more than 1 target); therefore we filter out
+    # direct dep components that have any intersection with the input component to avoid
+    # a rule cycle.
+    # TODO: Consider adding some further sanity checks to ensure that if there is any intersection,
+    #   then it is actually a full equality of components rather than a partial overlap.
+    #   In theory this should be guaranteed by the coarsening logic.
```

Comment:
As above: merging/deduplicating of the dependencies of the set should likely be baked into `CoarsenedTargets`. Will do as part of #12251.